Fix attaching to USDT probes with semaphore and USDT runtime tests

dalehamel opened the pull request:
This should fix #612 which @anelson correctly identified the issue in his detailed bug report.

I erroneously added code to close the USDT context in an inappropriate spot, which broke the semaphore handling functionality.

I've added a runtime test that will prevent this issue in the future, as well as fixed an issue with runtime tests for USDT that was introduced in #610 where the macro wasn't actually defined, which caused runtime tests to fail even on platforms with `sys/sdt.h` (cc @jasonk000)

@ajor can you take a look please?

dalehamel commented:
Just noticed that @mmarchini has a draft PR (#690) that fixes the issue from #610. i'll adopt his code from there to avoid a conflict

dalehamel commented on tests/testprogs/usdt_semaphore_test.c:
```diff
@@ -0,0 +1,34 @@
+#define _SDT_HAS_SEMAPHORES 1
```
I made this a separate test program, as this macro forces *all* probes to have a semaphore, and we want to be able to test both semaphore and non-semaphore cases.

danobi commented:
I merged the other PR. Could you please rebase?

mmarchini commented:
I'm ok with landing this, but I think we should open a follow-up issue to properly close `ctx`.

dalehamel commented:
@mmarchini I agree. I’ll rebase, file the issue, and get started on fixing the dangling context. I have an idea of how to properly solve that properly.

javierhonduco commented:
LGTM! Thanks for fixing this @dalehamel, we also hit this problem with our Python probes (and nice to see you around btw, long time no see! 😀)

dalehamel commented:
@mmarchini I've rebased, fixed-up the tests, and tried to make the runtime tests here follow your established idioms from #690.

I've filed #744 to properly deal with the USDT context, I'll take a look at that issue soon.
